From 40c836713bbebed6d38e4fa07a0fefe2f5dfe639 Mon Sep 17 00:00:00 2001
From: denglei <denglei@eswincomputing.com>
Date: Fri, 29 Aug 2025 09:00:39 +0800
Subject: [PATCH 492/541] fix:Resolve NPU driver bugs.

Changelogs:
	1.Resolve reset deadlock and PM counting errors.
	2.Resolve the issue of no corresponding dsp operator
	  driver crash when submitting the model.

Signed-off-by: denglei <denglei@eswincomputing.com>
---
 drivers/soc/eswin/ai_driver/npu/nvdla_hw.c  | 157 +++++++++-----------
 drivers/soc/eswin/ai_driver/npu/parse_dep.c |   8 +
 2 files changed, 77 insertions(+), 88 deletions(-)

diff --git a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
index 1a11a498258d..132cf3ab3583 100644
--- a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
+++ b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
@@ -507,7 +507,6 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 	int is_reset = 0;
 	struct nvdla_device *nvdla_dev = dev_get_drvdata(d);
 	struct win_engine *engine = npu_get_win_engine(nvdla_dev);
-
 	int ret, i;
 
 	if (count == 0 || count > 128) {
@@ -516,110 +515,92 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 	}
 
 	memcpy(buf_in, buf, count);
-
 	ptr = strstr(buf_in, "reset");
 	if (ptr != NULL) {
 		is_reset = 1;
 	}
 
 	ptr = strstr(buf_in, "npu");
-	mutex_lock(&engine->reset_mutex);
-	if (ptr != NULL) {
-		if (is_reset == 1) {
-			ret = npu_pm_get(nvdla_dev);
-			if (ret < 0) {
-				dla_error("%s, %d, pm get sync err, ret=%d.\n", __func__,
-					__LINE__, ret);
-				return ret;
-			}
-			engine->engine_is_alive = false;
-			npu_drop_all_frame(nvdla_dev, false);
-			memset(engine->host_node, 0, sizeof(host_node_t));
-			memset(nvdla_dev->emission_base, 0,
-			       E31_EMISSION_DTIM_SIZE);
-			memset(nvdla_dev->program_base, 0,
-			       E31_PROGRAM_DTIM_SIZE);
-
-			engine->tiktok = 0;
-
-			edma_free(nvdla_dev);
-
-			ret = npu_hardware_reset(nvdla_dev);
-			if (ret) {
-				dla_error("hardware reset err, ret=%d.\n", ret);
-				goto exit;
-			}
+	if (ptr != NULL && is_reset == 1) {
+		mutex_lock(&engine->reset_mutex);
+		ret = npu_pm_get(nvdla_dev);
+		if (ret < 0) {
+			dla_error("%s, %d, pm get sync err, ret=%d.\n", __func__,
+				__LINE__, ret);
+			mutex_unlock(&engine->reset_mutex);
+			return count;
+		}
+		engine->engine_is_alive = false;
+		npu_drop_all_frame(nvdla_dev, false);
+		memset(engine->host_node, 0, sizeof(host_node_t));
+		memset(nvdla_dev->emission_base, 0,	E31_EMISSION_DTIM_SIZE);
+		memset(nvdla_dev->program_base, 0, E31_PROGRAM_DTIM_SIZE);
 
-			npu_tbu_power(&nvdla_dev->pdev->dev, true);
-
-			clk_prepare_enable(nvdla_dev->mbox_pclk);
-			clk_prepare_enable(nvdla_dev->mbox_pclk_device);
-			reset_control_reset(nvdla_dev->mbox_rst_device);
-			reset_control_reset(nvdla_dev->mbox_rst);
-			writel(0x1,
-			       nvdla_dev->mbox_rx_base + MBOX_NPU_INT_OFFSET);
-			writel(ESWIN_MAILBOX_NPU_LOCK_BIT,
-			       nvdla_dev->mbox_rx_base + MBOX_NPU_WR_LOCK);
-			npu_restart_init(nvdla_dev);
-
-			host_ipc_initialize(
-				(u64)((struct win_engine
-					       *)(nvdla_dev->win_engine))
-					->host_node,
-				(u32)((struct win_engine
-					       *)(nvdla_dev->win_engine))
-					->host_node_iova,
-				(u64)nvdla_dev->emission_base,
-				(u64)nvdla_dev->program_base);
-
-			reset_uart_mutex(nvdla_dev);
-			for (i = 0; i < NUM_NPU_CORES; i++) {
-				/* initialize the node id for all e31 */
-				io_write((u8 *)nvdla_dev->e31_mmio_base +
-						 NPU_CPU_OFFSET +
-						 NPU_CPU_SIZE * i +
-						 NPU_DTIM_OFFSET +
-						 ADDR_OFFSET(cpu_node_t,
-							     node_id),
-					 i);
-				activate_system(nvdla_dev->e31_mmio_base, i);
-				if (i == 0) {
-					msleep(3);
-				}
+		engine->tiktok = 0;
+		edma_free(nvdla_dev);
+
+		ret = npu_hardware_reset(nvdla_dev);
+		if (ret) {
+			dla_error("hardware reset err, ret=%d.\n", ret);
+			goto exit;
+		}
+
+		npu_tbu_power(&nvdla_dev->pdev->dev, true);
+
+		clk_prepare_enable(nvdla_dev->mbox_pclk);
+		clk_prepare_enable(nvdla_dev->mbox_pclk_device);
+		reset_control_reset(nvdla_dev->mbox_rst_device);
+		reset_control_reset(nvdla_dev->mbox_rst);
+		writel(0x1, nvdla_dev->mbox_rx_base + MBOX_NPU_INT_OFFSET);
+		writel(ESWIN_MAILBOX_NPU_LOCK_BIT, nvdla_dev->mbox_rx_base + MBOX_NPU_WR_LOCK);
+		npu_restart_init(nvdla_dev);
+
+		host_ipc_initialize(
+			(u64)((struct win_engine *)(nvdla_dev->win_engine))->host_node,
+			(u32)((struct win_engine *)(nvdla_dev->win_engine))->host_node_iova,
+			(u64)nvdla_dev->emission_base, (u64)nvdla_dev->program_base);
+
+		reset_uart_mutex(nvdla_dev);
+		for (i = 0; i < NUM_NPU_CORES; i++) {
+			/* initialize the node id for all e31 */
+			io_write((u8 *)nvdla_dev->e31_mmio_base +
+					 NPU_CPU_OFFSET + NPU_CPU_SIZE * i + NPU_DTIM_OFFSET +
+					 ADDR_OFFSET(cpu_node_t, node_id), i);
+			activate_system(nvdla_dev->e31_mmio_base, i);
+			if (i == 0) {
+				msleep(3);
 			}
+		}
 
+		ret = check_system_activated(nvdla_dev->e31_mmio_base);
+		if (ret == false) {
 			ret = check_system_activated(nvdla_dev->e31_mmio_base);
-			if (ret == false) {
-				ret = check_system_activated(
-					nvdla_dev->e31_mmio_base);
-			}
-			if (nvdla_dev->e31_fw_virt_base) {
-				iommu_unmap_rsv_iova(
-					d, nvdla_dev->e31_fw_virt_base,
-					nvdla_dev->e31_nim_iova,
-					nvdla_dev->e31_fw_size);
-			}
-			if (ret == false) {
-				dla_error("e31 not bootup, error.\n");
-				goto exit;
-			}
+		}
+		if (nvdla_dev->e31_fw_virt_base) {
+			iommu_unmap_rsv_iova(d, nvdla_dev->e31_fw_virt_base,
+								 nvdla_dev->e31_nim_iova,
+								 nvdla_dev->e31_fw_size);
+		}
+		if (ret == false) {
+			dla_error("e31 not bootup, error.\n");
+			goto exit;
+		}
 
-			ret = edma_init(nvdla_dev);
-			if (ret) {
-				dla_error("edma_init fail\n");
-				goto exit;
-			}
-			engine->engine_is_alive = true;
-			npu_pm_put(nvdla_dev);
-		} else {
-			dla_error("err:input err.\n");
+		ret = edma_init(nvdla_dev);
+		if (ret) {
+			dla_error("edma_init fail\n");
+			goto exit;
 		}
+		engine->engine_is_alive = true;
+	} else {
+		dla_error("err:input err.\n");
+		return count;
 	}
 
 exit:
 	npu_clk_reset_print(nvdla_dev->pdev, nvdla_dev->numa_id);
-	mutex_unlock(&engine->reset_mutex);
 	npu_pm_put(nvdla_dev);
+	mutex_unlock(&engine->reset_mutex);
 	return count;
 }
 
diff --git a/drivers/soc/eswin/ai_driver/npu/parse_dep.c b/drivers/soc/eswin/ai_driver/npu/parse_dep.c
index c097ef2af355..e21c41146346 100644
--- a/drivers/soc/eswin/ai_driver/npu/parse_dep.c
+++ b/drivers/soc/eswin/ai_driver/npu/parse_dep.c
@@ -98,6 +98,10 @@ static int config_sequence_setup(struct win_executor *executor, int op_num,
 						    &task->surface_desc[op_idx],
 						    executor->tensor_set[pcer],
 						    j);
+		if (ret < 0) {
+			dla_error("%s, %d, tensor unfold error.\n", __func__, __LINE__);
+			return ret;
+		}
 		if (pcer_interface->rdma_check != NULL) {
 			rdma = pcer_interface->rdma_check(
 				NULL, &task->op_desc[op_idx],
@@ -107,6 +111,10 @@ static int config_sequence_setup(struct win_executor *executor, int op_num,
 		ret = pcer_interface->prepare_prog_data(
 			executor, rdma, j, op_idx, &task->op_desc[op_idx],
 			&task->surface_desc[op_idx]);
+		if (ret < 0) {
+			dla_error("%s, %d, prepare prog data error.\n", __func__, __LINE__);
+			return ret;
+		}
 		op_type_pos[pcer]++;
 	}
 
-- 
2.54.0

